src/shelly_hap_rgb.cpp

```diff
@@ -101,113 +103,126 @@ Status RGB::Init() {
       true /* supports_notification */,
-      std::bind(&RGB::HandleSaturationWrite, this, _1, _2, _3),
+      std::bind(&RGBWLight::HandleSaturationWrite, this, _1, _2, _3),
       kHAPCharacteristicDebugDescription_Saturation);
   state_notify_chars_.push_back(saturation_char);
   AddChar(saturation_char);
 
   return Status::OK();
 }
 
-void RGB::HSVtoRGB(float h, float s, float v, float &r, float &g, float &b) {
-  if (s == 0.0) {
+void RGBWLight::HSVtoRGBW(const HSV &hsv, RGBW &rgbw) const {
+  if (hsv.s == 0.0) {
     // if saturation is zero than all rgb channels same as brightness
-    r = g = b = v;
-    return;
-  }
+    rgbw.r = rgbw.g = rgbw.b = hsv.v;
+  } else {
+    // otherwise calc rgb from hsv
+    int i = static_cast<int>(hsv.h * 6);
+    float f = (hsv.h * 6.0f - i);
+    float p = hsv.v * (1.0f - hsv.s);
+    float q = hsv.v * (1.0f - f * hsv.s);
+    float t = hsv.v * (1.0f - (1.0f - f) * hsv.s);
+
+    switch (i % 6) {
+      case 0:
+        rgbw.r = hsv.v;
+        rgbw.g = t;
+        rgbw.b = p;
+        break;
 
-  int i = static_cast<int>(h * 6);
-  float f = (h * 6.0f - i);
-  float p = v * (1.0f - s);
-  float q = v * (1.0f - f * s);
-  float t = v * (1.0f - (1.0f - f) * s);
-
-  switch (i % 6) {
-    case 0:
-      r = v;
-      g = t;
-      b = p;
-      break;
+      case 1:
+        rgbw.r = q;
+        rgbw.g = hsv.v;
+        rgbw.b = p;
+        break;
 
-    case 1:
-      r = q;
-      g = v;
-      b = p;
-      break;
+      case 2:
+        rgbw.r = p;
+        rgbw.g = hsv.v;
+        rgbw.b = t;
+        break;
 
-    case 2:
-      r = p;
-      g = v;
-      b = t;
-      break;
+      case 3:
+        rgbw.r = p;
+        rgbw.g = q;
+        rgbw.b = hsv.v;
+        break;
 
-    case 3:
-      r = p;
-      g = q;
-      b = v;
-      break;
+      case 4:
+        rgbw.r = t;
+        rgbw.g = p;
+        rgbw.b = hsv.v;
+        break;
 
-    case 4:
-      r = t;
-      g = p;
-      b = v;
-      break;
+      case 5:
+        rgbw.r = hsv.v;
+        rgbw.g = p;
+        rgbw.b = q;
+        break;
+    }
+  }
 
-    case 5:
-      r = v;
-      g = p;
-      b = q;
-      break;
+  if (GetMode() == Mode::kRgbw) {
+    // apply white channel to rgb if activated
+    rgbw.w = std::min(rgbw.r, std::min(rgbw.g, rgbw.b));
+    rgbw.r = rgbw.r - rgbw.w;
+    rgbw.g = rgbw.g - rgbw.w;
+    rgbw.b = rgbw.b - rgbw.w;
+  } else {
+    // otherwise turn white channel off
+    rgbw.w = 0.0f;
   }
 }
 
-void RGB::SetOutputState(const char *source) {
+void RGBWLight::SetOutputState(const char *source) {
   LOG(LL_INFO,
       ("state: %s, brightness: %i, hue: %i, saturation: %i", OnOff(cfg_->state),
        cfg_->brightness, cfg_->hue, cfg_->saturation));
 
-  float r = 0, g = 0, b = 0;
+  RGBW rgbw{0.0f, 0.0f, 0.0f, 0.0f};
 
   if (cfg_->state) {
-    float h = cfg_->hue / 360.0f;
-    float s = cfg_->saturation / 100.0f;
-    float v = cfg_->brightness / 100.0f;
+    HSV hsv;
```

Comment:
```c
HSV hsv = {
    .h = ...,
    .s = ...,
    .v = ...,
};
```
whether c or c++, always use designated initializers to initialize structs.
this guarantees that new members (if ever added) are initialized with 0.